### PR TITLE
Add unzip

### LINKED ISSUE #13942: Base.unzip()
Hi there,

apologies if this has already been addressed somewhere, but is there a reason that there is no `unzip()` function in Base?

Ideally this would be a function that would take a `Vector{Tuple{ ... }}` and return a `Tuple{Vector, ..., Vector}` for output. E.g.
```
julia> v = [(1,"a",:meow), (2,"b",:woof), (3,"c",:doh!)]; unzip(v)
([1,2,3],ASCIIString["a","b","c"],[:meow,:woof,:doh!])
```

A naive implementation might be something like
```julia
function unzip(input::Vector)
    n = length(input)
    types  = map(typeof, first(input))
    output = map(T->Vector{T}(n), types)

    for i = 1:n
       @inbounds for (j, x) in enumerate(input[i])
           (output[j])[i] = x
       end
    end
    return (output...)
end
```

### PR DESCRIPTION
**Pull request by bramtayl**
I finally switched to Ubuntu, and suddenly building and testing Base became much more manageable! I think this should be pretty close to good.

Closes #13942

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -2467,3 +2467,192 @@ end
...
+partial_map(call, fixed, variables1::Tuple{}, variables2::Tuple{}) = ()
+partial_map(call, fixed, variables1, variables2) =
+    call(fixed, first(variables1), first(variables2)),
+    partial_map(call, fixed, tail(variables1), tail(variables2))...
+
+function reduce_unrolled(call, item)
+    item
+end
+function reduce_unrolled(call, item1, item2, rest...)
+    reduce_unrolled(call, call(item1, item2), rest...)
```
Comment (tkf) [inline base/array.jl]: Isn't this equivalent to `Base.afoldl`?

File: base/array.jl [hunk 2]
```diff
@@ -2467,3 +2467,192 @@ end
 
 intersect(itr, itrs...) = _shrink(intersect!, itr, itrs)
 setdiff(  itr, itrs...) = _shrink(setdiff!, itr, itrs)
+
+map_unrolled(call, variables::Tuple{}) = ()
+map_unrolled(call, variables) =
+    call(first(variables)), map_unrolled(call, tail(variables))...
```
Comment (tkf) [inline base/array.jl]: I think `Base.map` does this already.

File: base/array.jl [hunk 3]
```diff
@@ -2467,3 +2467,192 @@ end
...
+
+map_unrolled(call, variables1::Tuple{}, variables2::Tuple{}) = ()
+map_unrolled(call, variables1, variables2) =
+    call(first(variables1), first(variables2)),
+    map_unrolled(call, tail(variables1), tail(variables2))...
+
+partial_map(call, fixed, variables::Tuple{}) = ()
+partial_map(call, fixed, variables) =
+    call(fixed, first(variables)), partial_map(call, fixed, tail(variables))...
```
Comment (tkf) [inline base/array.jl]: Didn't using a closure with map work?

File: base/array.jl (hunk 1 above)
Comment (bramtayl) [reply base/array.jl]: Yup, but I wanted to make super duper sure that it was going to unroll, not iterate

File: base/array.jl (hunk 2 above)
Comment (bramtayl) [reply base/array.jl]: Yup, but I wanted to make super duper sure that it was going to unroll, not iterate

File: base/array.jl (hunk 3 above)
Comment (bramtayl) [reply base/array.jl]: #31909

Comment (tkf) [reply base/array.jl]:
```juila
julia> partial_map(f, c, xs) = map(x -> f(c, x), xs)
partial_map (generic function with 1 method)

julia> demo() = partial_map(+, rand(Bool) ? 1 : missing, (1, 2))
demo (generic function with 1 method)

julia> @code_warntype demo()
Variables
  #self#::Core.Compiler.Const(demo, false)
  @_2::Union{Missing, Int64}

Body::Union{Tuple{Missing,Missing}, Tuple{Int64,Int64}}
1 ─ %1 = Main.rand(Main.Bool)::Bool
└──      goto #3 if not %1
2 ─      (@_2 = 1)
└──      goto #4
3 ─      (@_2 = Main.missing)
4 ┄ %6 = @_2::Union{Missing, Int64}
│   %7 = Core.tuple(1, 2)::Core.Compiler.Const((1, 2), false)
│   %8 = Main.partial_map(Main.:+, %6, %7)::Union{Tuple{Missing,Missing}, Tuple{Int64,Int64}}
└──      return %8
```

Comment (bramtayl) [reply base/array.jl]: As long as #31909 is open I'd like to avoid closures

**Comment by bramtayl** [comment]
CI failures unrelated, correct?

File: base/array.jl (hunk 1 above)
Comment (mschauer) [reply base/array.jl]: Let's add a comment: `Functionally equivalent to Base.afold, but...` for posterity

File: base/array.jl [hunk 4]
```diff
@@ -2467,3 +2467,192 @@ end
...
+end
+
+@propagate_inbounds function Rows{Row, Dimension}(columns::Columns) where {Row, Dimension, Columns}
+    @boundscheck if !same_axes(columns...)
+        throw(DimensionMismatch("All arguments to `Rows` must have the same axes"))
+    end
+    Rows{Row, Dimension, Columns}(columns)
+end
+
+get_model(columns) = first(columns)
```
Comment (mschauer) [inline base/array.jl]: can this have a more telling function name?

File: base/array.jl (hunk 1 above)
Comment (bramtayl) [reply base/array.jl]: Sure. I've at several times considering adding a sub-module of unrolled functions to base; perhaps the time has come?

File: base/array.jl (hunk 4 above)
Comment (bramtayl) [reply base/array.jl]: Sure. What should I call it?

Comment (mschauer) [reply base/array.jl]: I have not yet understood what this function does so this is on you ;-)

Comment (bramtayl) [reply base/array.jl]: The dimensions and axes of Rows are defined by its first column. If there are no columns, it defaults to `1:0` (the empty unit range). Does that make sense?

Comment (mschauer) [reply base/array.jl]: Maybe `dummy_row`?

Comment (bramtayl) [reply base/array.jl]: Sure

File: base/array.jl (hunk 2 above)
Comment (bramtayl) [reply base/array.jl]: I put the unrolled operations into a separate mini-module

File: base/array.jl (hunk 1 above)
Comment (bramtayl) [reply base/array.jl]: I put the unrolled operations into a separate mini-module

**Comment by bramtayl** [comment]
CI failures unrelated

**Comment by bramtayl** [comment]
@StefanKarpinski can you assign someone to review this?

**Comment by StefanKarpinski** [comment]
It feels like a _lot_ of stuff to implement `unzip`. Here's a version that I cooked up:
```jl
"""
    unzip(itrs) -> Vector{<:Vector}

The `unzip` function takes an iterator of iterators and returns a vector of
vectors such that `zip(unzip(itrs)...) == map(itr->tuple(itr...), itrs)`.

# Examples

    julia> unzip(enumerate("Hello"))
    2-element Array{Array{T,1} where T,1}:
     [1, 2, 3]
     ['a', 'b', 'c']

    julia> unzip([[1, 'a'], [2.5, 'z'], [0, 'x']])
    2-element Array{Array{T,1} where T,1}:
     Real[1, 2.5, 0]
     ['a', 'z', 'x']

"""
function unzip(itrs)
    outer = iterate(itrs)
    outer === nothing && return []
    vals, state = outer
    vecs = Vector[[val] for val in vals]
    local n
    if Base.haslength(itrs)
        n = length(itrs)
        for vec in vecs
            sizehint!(vec, n)
        end
    end
    while true
        outer = iterate(itrs, state)
        outer === nothing && return vecs
        vals, state = outer
        for (j, val) = enumerate(vals)
            vec = vecs[j]
            if !(val isa eltype(vec))
                T = Base.promote_typejoin(typeof(val), eltype(vec))
                vec = vecs[j] = copyto!(similar(vec, T), vec)
                @isdefined(n) && sizehint!(vec, n)
            end
            push!(vec, val)
        end
    end
end
```

**Comment by StefanKarpinski** [comment]
I guess the question is what one wants `unzip` to do. Here's a simpler version that allows unzipping unevenly sized arrays—which is of questionable desirability—but makes it simpler to implement:
```jl
function unzip(itrs; uneven::Bool=false)
    n = Base.haslength(itrs) ? length(itrs) : -1
    vecs = Vector[]
    for itr in itrs
        for (j, x) in enumerate(itr)
            if length(vecs) < j
                v = [x]
                push!(vecs, v)
                n ≥ 0 && sizehint!(v, n)
            else
                v = vecs[j]
                if !(x isa eltype(v))
                    T = Base.promote_typejoin(typeof(x), eltype(v))
                    v = vecs[j] = copyto!(similar(v, T), v)
                    n ≥ 0 && sizehint!(v, n)
                end
                push!(v, x)
            end
        end
        uneven || length(first(vecs)) == length(last(vecs)) ||
            throw(ArgumentError("unzip called with uneven iterators"))
    end
    return vecs
end
```
Works the same on evenly sized arrays but here's an example with unevenly sized ones:
```jl
julia> unzip([[1, 'a'], [2.5, 'z', "extra"], [0, 'x']], uneven=true)
3-element Array{Array{T,1} where T,1}:
 Real[1, 2.5, 0]
 ['a', 'z', 'x']
 ["extra"]
```
This just goes through each iterable thing and collects into the vector at the corresponding index.

**Pull request closed without merging**